File: backend/src/main/java/com/hehe/thesocial/service/metadata/love/LoveService.java
package com.hehe.thesocial.service.metadata.love;

import com.hehe.thesocial.dto.response.metadata.LoveResponse;

public interface LoveService {
    /**
     * Thêm userDetailId vào danh sách lovedBy của feedItem
     * Tăng loveCount trong metadata
     * @param feedItemId ID của feed item
     * @param userDetailId ID của user detail
     * @return LoveResponse chứa trạng thái loved và loveCount mới
     */
    LoveResponse addLove(String feedItemId, String userDetailId);

    /**
     * Xóa userDetailId khỏi danh sách lovedBy của feedItem
     * Giảm loveCount trong metadata
     * @param feedItemId ID của feed item
     * @param userDetailId ID của user detail
     * @return LoveResponse chứa trạng thái loved và loveCount mới
     */
    LoveResponse removeLove(String feedItemId, String userDetailId);

    /**
     * Kiểm tra xem userDetail đã love feedItem chưa
     * @param feedItemId ID của feed item
     * @param userDetailId ID của user detail
     * @return LoveResponse chứa trạng thái loved và loveCount hiện tại
     */
    LoveResponse checkLoveStatus(String feedItemId, String userDetailId);
}

File: mobile-app/src/Services/LoveService.ts
import { api } from "./HttpClient";

export interface LoveResponse {
    loved: boolean;
    loveCount: number;
}

export interface ApiResponse<T> {
    result: T;
    message: string;
}

export const LikeService = {
    /**
     * Thêm love cho feed item
     */
    addLove: async (feedItemId: string): Promise<LoveResponse> => {
        try {
            console.log(`[LikeService] ========== addLove called ==========`);
            console.log(`[LikeService] FeedItem ID:`, feedItemId);
            console.log(`[LikeService] ID type:`, typeof feedItemId);
            console.log(`[LikeService] ID length:`, feedItemId?.length);
            console.log(`[LikeService] Calling API: POST /feed-items/${feedItemId}/love`);

            const response = await api.post<ApiResponse<LoveResponse>>(
                `/feed-items/${feedItemId}/love`
            );

            console.log("[LikeService] ✅ Love added successfully:", response.data);
            return response.data.result;
        } catch (error: any) {
            console.error("[LikeService] ❌ Error adding love:", {
                feedItemId,
                message: error.message,
                response: error.response?.data,
                status: error.response?.status,
                url: error.config?.url,
            });
            throw error;
        }
    },

    /**
     * Xóa love khỏi feed item
     */
    removeLove: async (feedItemId: string): Promise<LoveResponse> => {
        try {
            console.log(`[LikeService] ========== removeLove called ==========`);
            console.log(`[LikeService] FeedItem ID:`, feedItemId);
            console.log(`[LikeService] Calling API: DELETE /feed-items/${feedItemId}/love`);

            const response = await api.delete<ApiResponse<LoveResponse>>(
                `/feed-items/${feedItemId}/love`
            );

            console.log("[LikeService] ✅ Love removed successfully:", response.data);
            return response.data.result;
        } catch (error: any) {
            console.error("[LikeService] ❌ Error removing love:", {
                feedItemId,
                message: error.message,
                response: error.response?.data,
                status: error.response?.status,
                url: error.config?.url,
            });
            throw error;
        }
    },

    /**
     * Kiểm tra trạng thái love
     */
    checkLoveStatus: async (feedItemId: string): Promise<LoveResponse> => {
        try {
            console.log(`[LikeService] Checking love status for feedItem: ${feedItemId}`);
            const response = await api.get<ApiResponse<LoveResponse>>(
                `/feed-items/${feedItemId}/love`
            );

            console.log("[LikeService] Love status checked:", response.data);
            return response.data.result;
        } catch (error: any) {
            console.error("[LikeService] Error checking love status:", {
                message: error.message,
                response: error.response?.data,
                status: error.response?.status,
            });
            throw error;
        }
    },
};